lms/djangoapps/grades/events.py

```diff
@@ -0,0 +1,134 @@
+from crum import get_current_user
+from eventtracking import tracker
+from track import contexts
+from track.event_transaction_utils import (
+    create_new_event_transaction_id,
+    get_event_transaction_id,
+    get_event_transaction_type,
+    set_event_transaction_type
+)
+
+
+COURSE_GRADE_CALCULATED = u'edx.grades.course.grade_calculated'
+GRADES_OVERRIDE_EVENT_TYPE = u'edx.grades.problem.score_overridden'
+GRADES_RESCORE_EVENT_TYPE = u'edx.grades.problem.rescored'
+PROBLEM_SUBMITTED_EVENT_TYPE = u'edx.grades.problem.submitted'
+STATE_DELETED_EVENT_TYPE = u'edx.grades.problem.state_deleted'
+SUBSECTION_OVERRIDE_EVENT_TYPE = u'edx.grades.subsection.score_overridden'
+SUBSECTION_GRADE_CALCULATED = u'edx.grades.subsection.grade_calculated'
+
+
+def grade_updated(**kwargs):
+    """
+    Emits the appropriate grade-related event after checking for which
+    event-transaction is active.
+
+    Emits a problem.submitted event only if there is no current event
+    transaction type, i.e. we have not reached this point in the code via
+    an outer event type (such as problem.rescored or score_overridden).
+    """
+    root_type = get_event_transaction_type()
+
+    if not root_type:
+        root_id = get_event_transaction_id()
+        if not root_id:
+            root_id = create_new_event_transaction_id()
+        set_event_transaction_type(PROBLEM_SUBMITTED_EVENT_TYPE)
+        tracker.emit(
+            unicode(PROBLEM_SUBMITTED_EVENT_TYPE),
+            {
+                'user_id': unicode(kwargs['user_id']),
+                'course_id': unicode(kwargs['course_id']),
+                'problem_id': unicode(kwargs['usage_id']),
+                'event_transaction_id': unicode(root_id),
+                'event_transaction_type': unicode(PROBLEM_SUBMITTED_EVENT_TYPE),
+                'weighted_earned': kwargs.get('weighted_earned'),
+                'weighted_possible': kwargs.get('weighted_possible'),
+            }
+        )
+
+    elif root_type in [GRADES_RESCORE_EVENT_TYPE, GRADES_OVERRIDE_EVENT_TYPE]:
+        current_user = get_current_user()
+        instructor_id = getattr(current_user, 'id', None)
+        tracker.emit(
+            unicode(root_type),
+            {
+                'course_id': unicode(kwargs['course_id']),
+                'user_id': unicode(kwargs['user_id']),
+                'problem_id': unicode(kwargs['usage_id']),
+                'new_weighted_earned': kwargs.get('weighted_earned'),
+                'new_weighted_possible': kwargs.get('weighted_possible'),
+                'only_if_higher': kwargs.get('only_if_higher'),
+                'instructor_id': unicode(instructor_id),
+                'event_transaction_id': unicode(get_event_transaction_id()),
+                'event_transaction_type': unicode(root_type),
+            }
+        )
+
+    elif root_type in [SUBSECTION_OVERRIDE_EVENT_TYPE]:
```

Comment:
I know we didn't write this but is there a stylistic or functional reason to favor this syntax over `root_type == SUBSECTION_OVERRIDE_EVENT_TYPE`?

Reply:
not from my POV - unless the writer felt that this condition would eventually evolve over time to include additional event-types in the list.